test/functional/p2p_orphan_handling.py

```diff
@@ -0,0 +1,439 @@
+        peer_txrequest.send_and_ping(msg_inv([CInv(t=MSG_WTX, h=int(inflight_parent_AB["tx"].getwtxid(), 16))]))
+        self.fastforward(NONPREF_PEER_TX_DELAY)
+        peer_txrequest.wait_for_getdata([int(inflight_parent_AB["tx"].getwtxid(), 16)])
+
+        self.log.info("The node should not request a parent if it has an in-flight txrequest")
+        # Relay orphan child_A
+        self.relay_transaction(peer_orphans, child_A["tx"])
+        self.fastforward(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        # Both missing parents should be requested.
+        peer_orphans.wait_for_parent_requests([int(missing_parent_A["txid"], 16), int(missing_parent_AB["txid"], 16)])
+
+        self.log.info("The node should not request a parent if it has an in-flight orphan parent request")
+        # Relay orphan child_B
+        self.relay_transaction(peer_orphans, child_B["tx"])
+        self.fastforward(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        # Only missing_parent_B should be requested. Not inflight_parent_AB or missing_parent_AB
+        # because they are already being requested from peer_orphans.
+        peer_orphans.wait_for_parent_requests([int(missing_parent_B["txid"], 16)])
+        peer_orphans.assert_never_requested(int(inflight_parent_AB["txid"], 16))
+
+    @cleanup
+    def test_orphan_of_orphan(self):
+        node = self.nodes[0]
+        peer = node.add_p2p_connection(PeerTxRelayer())
+
+        self.log.info("Test handling of an orphan with a parent who is another orphan")
+        missing_grandparent = self.wallet_nonsegwit.create_self_transfer()
+        missing_parent_orphan = self.wallet_nonsegwit.create_self_transfer(utxo_to_spend=missing_grandparent["new_utxo"])
+        missing_parent = self.wallet_nonsegwit.create_self_transfer()
+        orphan = self.wallet_nonsegwit.create_self_transfer_multi(utxos_to_spend=[missing_parent["new_utxo"], missing_parent_orphan["new_utxo"]])
+
+        # The node should put missing_parent_orphan into the orphanage and request missing_grandparent
+        self.relay_transaction(peer, missing_parent_orphan["tx"])
+        self.fastforward(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        peer.wait_for_parent_requests([int(missing_grandparent["txid"], 16)])
+
+        # The node should put the orphan into the orphanage and request missing_parent, skipping
+        # missing_parent_orphan because it already has it in the orphanage.
+        self.relay_transaction(peer, orphan["tx"])
+        self.fastforward(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        peer.wait_for_parent_requests([int(missing_parent["txid"], 16)])
+
+    @cleanup
+    def test_orphan_inherit_rejection(self):
+        node = self.nodes[0]
+        peer1 = node.add_p2p_connection(PeerTxRelayer())
+        peer2 = node.add_p2p_connection(PeerTxRelayer())
+        peer3 = node.add_p2p_connection(PeerTxRelayer())
+
+        self.log.info("Test orphan handling when nonsegwit parent paid 0 fee")
```

Comment:
bit of a repeat with test_orphan_rejected_parents_exceptions ?

Reply:
(I didn't change it but I also don't feel strongly, could combine if you want)

Reply:
nope it's fine